### ISSUE TITLE
Player (OSX High Sierra)- freezes during calibration in debug mode in Pycharm and crashes in VS Code

### ISSUE THREAD

**Issue by performlabrit** [pupil-labs/pupil#1029]
A strange one, I know.  Without fail, my build stalls during calibration when run in debug mode in Pycharm.  This is a repeatable issue. I have re-forked and have the same issue.  I have tried erasing the offline data folder.  any other suggestions?  

This is the log output prior to the halting.
player - [INFO] gaze_producers: Calibrating "1" in 3d mode...

Running on a Mac w/High Sierra, and the latest build of pycharm.

**Comment by performlabrit** [comment]
Update:  Learned how to use VSCode.  Similar issues, however, instead of freezing, Visual Studio code crashes.  Output below.  Normal run mode works just fine in both IDE, and this persists even with a fresh fork from github.

    from . screen_marker_calibration import Screen_Marker_Calibration
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_4/pupil_src/shared_modules/calibration_routines/screen_marker_calibration.py", line 28, in <module>
    from . finish_calibration import finish_calibration
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_4/pupil_src/shared_modules/calibration_routines/finish_calibration.py", line 19, in <module>
    from . optimization_calibration import bundle_adjust_calibration
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_4/pupil_src/shared_modules/calibration_routines/optimization_calibration/__init__.py", line 17, in <module>
    build_cpp_extension()
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_4/pupil_src/shared_modules/calibration_routines/optimization_calibration/build.py", line 25, in build_cpp_extension
    ret = sp.check_output(build_cmd).decode(sys.stdout.encoding)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 336, in check_output
    **kwargs).stdout
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/subprocess.py", line 418, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['/usr/local/opt/python3/bin/python3.6', 'setup.py', 'install', '--install-lib=/Users/gjdiaz/PycharmProjects/pupil_pl_4/pupil_src/shared_modules']' returned non-zero exit status 1.
player - [INFO] launchables.player: Process shutting down.

**Comment by performlabrit** [comment]
Last update.  Debug mode works in the Wing IDE.  Third IDE is the charm, eh?  Not an idea solution.

**Comment by performlabrit** [comment]
I've tested, and It's unfortunate that this problem is independent of the video.  I will now install the source code on my friend's mac and see if it is just my machine.

**Comment by performlabrit** [comment]
Still crashing.  Here is output from a run via console using PDB.

gbook:pupil_src gjdiaz$ python3 -m pdb main.py player
> /Users/gjdiaz/PycharmProjects/pupil_pl_2/pupil_src/main.py(10)<module>()
-> '''
(Pdb) cont
MainProcess - [INFO] os_utils: Disabled idle sleep.
MainProcess - [INFO] os_utils: Re-enabled idle sleep.
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/pdb.py", line 1667, in main
    pdb._runscript(mainpyfile)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/pdb.py", line 1548, in _runscript
    self.run(statement)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/bdb.py", line 431, in run
    exec(cmd, globals, locals)
  File "<string>", line 1, in <module>
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_2/pupil_src/main.py", line 10, in <module>
    '''
  File "/Users/gjdiaz/PycharmProjects/pupil_pl_2/pupil_src/main.py", line 296, in launcher
    app_version,
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/process.py", line 105, in start
    self._popen = self._Popen(self)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/context.py", line 284, in _Popen
    return Popen(process_obj)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_fork.py", line 26, in __init__
    self._launch(process_obj)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/popen_spawn_posix.py", line 42, in _launch
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "/usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/spawn.py", line 172, in get_preparation_data
    main_mod_name = getattr(main_module.__spec__, "name", None)
AttributeError: module '__main__' has no attribute '__spec__'
Uncaught exception. Entering post mortem debugging
Running 'cont' or 'step' will restart the program
> /usr/local/Cellar/python3/3.6.4_1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/multiprocessing/spawn.py(172)get_preparation_data()
-> main_mod_name = getattr(main_module.__spec__, "name", None)
(Pdb)

**Comment by performlabrit** [comment]
Thank you for the suggestion.  Unfortunately, this causes an immediate crash.  I've set the spawn method back to spawn.   Output below.

```
MainProcess - [INFO] os_utils: Disabled idle sleep.
objc[36753]: +[NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called.
objc[36753]: +[NSPlaceholderDictionary initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
MainProcess - [INFO] os_utils: Re-enabled idle sleep.

```

**Comment by papr** [comment]
Your issue seems to be specific to `pdb` and `multiprocessing/spawn.py:172` calling `sys.modules['__main__'].__spec__`.

```python
import sys
# sys.modules['__main__'].__spec__ = None
print(sys.modules['__main__'].__spec__)
```

Running the above script with `python3.6 -m pdb <script>` crashes with the same exception as posted by you above. Running the the script without `-m pdb` simply prints `None`. Therefore this seems to be an issue related to a combination of pdb/spawn.

A custom/temporary work around is to add `sys.modules['__main__'].__spec__ = None` in https://github.com/pupil-labs/pupil/blob/master/pupil_src/main.py#L317

**Comment by performlabrit** [comment]
I can verify that this addition does fix debugging with pdb.  So, that's a big step in the right direction.  Unfortunately, Visual studio code and Pycharm seem to be having unique issues with their debuggers.  Again, this was not the case with previous commits.  This may be a coincidence. 

I've tried erasing all files in pupil_capture_settings and pupil_player_settings, and within the recording folder's offline data directory.  Any other cached data I should try removing?

**Comment by performlabrit** [comment]
An update:  I've just tried to debug on another Mac with a fresh fork, and fresh Python install.  Same issue with the debugger in Pycharm:.  Stuck on:   player - [INFO] gaze_producers: Calibrating "1" in 3d mode.  This strongly suggests that this issue is not particular to my machine, but to... all Macs, maybe?

**Comment by performlabrit** [comment]
Oh, and the fix posted above by paps does get sublime text running with a pdb debugger, but the debugger fails to stop at breakpoints inserted in plugins.  I don't think it likes the threading/processing.

**Comment by papr** [comment]
Should be fixed with `v1.9` https://github.com/pupil-labs/pupil/commit/cc681842615dbc170415afabf8ea909d2c6d0d33
